### PR TITLE
Inject defaults for certain attributes

### LINKED ISSUE #223: "Incorrect attribute value type"
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions
cdktf: 0.0.12
typescript: Version 3.9.7

### Affected Resource(s)

cdk deploy

### Debug Output

https://gist.github.com/pmartindev/a5a741043975855a2a1a6fcacf4f549a

### Expected Behavior

I would expect the file to create default/null values for the cdk.json and the cdk deploy  to succeed

### Actual Behavior

Executing `cdk deploy` produces the above error.

### Steps to Reproduce

1. Create an AWS security group construct (see: https://gist.github.com/pmartindev/b1ab0aa678678b11aadc30a3b23859d0)
2. run the `cdk deploy` command
3. Check output to see that the error is shown, and the cdk.tf.json does not contain the values that the error message specifies (see: https://gist.github.com/pmartindev/c9e93ffa5c7fa8c939b07252c9e90bcf)

### Important Factoids

This appears to be an issue with the Terraform .json files. A workaround is to set the values to null, however, this is not possible with Typescript. 

UPDATE: Adding the missing attributes with null values in the cdk.tf.json allows the terraform file to run correctly. 

https://gist.github.com/pmartindev/314d083116776634ee4b4a5f07f40c95

### References

https://github.com/terraform-providers/terraform-provider-aws/issues/8786

* #8786

### LINKED ISSUE #234: Inject Defaults during Code Generation
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

There's at least one occasion where Terraform JSON isn't fully compatible with HCL. https://github.com/terraform-providers/terraform-provider-aws/issues/8786#issuecomment-496935442 - It's some cases expected to provide explicit `null` values for optional attributes, e.g:

<details><summary>A Security Group example by @pmartindev</summary><p>

```ts
 const ecstasksg = new SecurityGroup(this, 'ecstasksg', {
      name: "tf-ecs-task-sg",
      description: "Allows traffic to the ALB.", 
      ingress: [{
        protocol: "tcp",
        fromPort: 80,
        toPort: 80,
        cidrBlocks: ["0.0.0.0/0"]
      },{
        protocol: "tcp",
        fromPort: 22,
        toPort: 22,
        cidrBlocks: ["0.0.0.0/0"]
      },{
        protocol: "tcp",
        fromPort: 443,
        toPort: 443,
        cidrBlocks: ["0.0.0.0/0"]
      }],
      egress: [{
        protocol: "-1",
        fromPort: 0,
        toPort: 0,
        cidrBlocks: ["0.0.0.0/0"]
      }],
    });

    // Ingress
    ecstasksg.addOverride("ingress.0.description", null);
    ecstasksg.addOverride("ingress.0.ipv6_cidr_blocks", null);
    ecstasksg.addOverride("ingress.0.prefix_list_ids", null);
    ecstasksg.addOverride("ingress.0.security_groups", null);
    ecstasksg.addOverride("ingress.0.selfAttribute", null);
    ecstasksg.addOverride("ingress.0.self", null);
    ecstasksg.addOverride("ingress.1.description", null);
    ecstasksg.addOverride("ingress.1.ipv6_cidr_blocks", null);
    ecstasksg.addOverride("ingress.1.prefix_list_ids", null);
    ecstasksg.addOverride("ingress.1.security_groups", null);
    ecstasksg.addOverride("ingress.1.selfAttribute", null);
    ecstasksg.addOverride("ingress.1.self", null);
    ecstasksg.addOverride("ingress.2.description", null);
    ecstasksg.addOverride("ingress.2.ipv6_cidr_blocks", null);
    ecstasksg.addOverride("ingress.2.prefix_list_ids", null);
    ecstasksg.addOverride("ingress.2.security_groups", null);
    ecstasksg.addOverride("ingress.2.selfAttribute", null);
    ecstasksg.addOverride("ingress.2.self", null);

    // Egress
    ecstasksg.addOverride("egress.0.description", null);
    ecstasksg.addOverride("egress.0.ipv6_cidr_blocks", null);
    ecstasksg.addOverride("egress.0.prefix_list_ids", null);
    ecstasksg.addOverride("egress.0.security_groups", null);
    ecstasksg.addOverride("egress.0.selfAttribute", null);
    ecstasksg.addOverride("egress.0.self", null);
```

</p></details>

While the need for `addOverride` could go away with https://github.com/hashicorp/terraform-cdk/issues/233, it'd still be required to explicitly specify the `null` values.

According to the [Terraform issue](https://github.com/terraform-providers/terraform-provider-aws/issues/8786#issuecomment-496935442) these resources are affected (from last year, would make sense to get a current list):

- `aws_default_route_table` resource route configuration block
- `aws_emr_cluster` resource step configuration block and nested hadoop_jar_step configuration block
- `aws_network_acl` resource ingress and egress configuration blocks
- `aws_route_table` resource route configuration block
- `aws_security_group` resource ingress and egress configuration blocks

To improve the usability, there should be a way to automatically provide these `null` values as default. In the best case, during code generation.

### References

- https://github.com/terraform-providers/terraform-provider-aws/issues/8786 
- https://github.com/hashicorp/terraform-cdk/issues/223
- https://github.com/hashicorp/terraform-cdk/issues/233

### LINKED ISSUE #443: KeyVaultAccessPolicy - Inappropriate value for attribute "access_policy": element 0: attributes "application_id"
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

cdktf => v 0.0.18
terraform => v 0.13.5
azurerm => 2.37.0
python => 3.8.6

### Affected Resource(s)

KeyVaultAccessPolicy

### Debug Output

https://gist.github.com/Mejato/11d8814f61bb7dc3237e1224fb388fd9

### Expected Behavior

cdktf should have constructed the appropriate policy based on the documentation.

imports.azurerm.KeyVaultAccessPolicy def __init__(self,
             *,
             application_id: **Optional**[str] = None,
             certificate_permissions: **Optional**[List[str]] = None,
             key_permissions: Optional[List[str]] = None,
             object_id: Optional[str] = None,
             secret_permissions: **Optional**[List[str]] = None,
             storage_permissions: **Optional**[List[str]] = None,
             tenant_id: Optional[str] = None) -> None

### Actual Behavior

Terraform planning stage fails - see debug output gist.

### Steps to Reproduce
azure_key_vault_access_policy = KeyVaultAccessPolicy(tenant_id=azure_ad_tenant_id, secret_permissions=['get', 'set'], object_id=azure_principle)

### Important Factoids

N/A

### PR DESCRIPTION
**Pull request by jsteinich**
Fixes #234
Fixes #223
Fixes #443

Simple default values injected during code generation based on fully qualified attribute name.
Currently not allowing users to specify their own defaults, but could expose through `cdktf.json` if that was desired.
Specifying individual attributes rather than whole block for a bit more flexibility, but could switch that.

### CODE REVIEW

**Comment by jsteinich** [comment]
@skorfmann Should I go ahead and other other known resources in the same manner?

**Comment by skorfmann** [comment]
> Currently not allowing users to specify their own defaults, but could expose through `cdktf.json` if that was desired.

I think that's mostly a stop-gap solution, so not sure what users would need to inject here. It's certainly good enough for a first iteration.

> Specifying individual attributes rather than whole block for a bit more flexibility, but could switch that.
> 

Makes sense 👍 

> Should I go ahead and other other known resources in the same manner?

Yes, please :)

File: packages/cdktf-cli/lib/get/generator/custom-defaults.ts [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    "azurerm.function_app.scm_ip_restriction.ip_address": null,
+    "azurerm.function_app.scm_ip_restriction.service_tag": null,
+    "azurerm.function_app.scm_ip_restriction.subnet_id": null,
+    "azurerm.function_app.scm_ip_restriction.virtual_network_subnet_id": null,
+    "azurerm.function_app.scm_ip_restriction.name": null,
+    "azurerm.function_app.scm_ip_restriction.priority": null,
+    "azurerm.function_app.scm_ip_restriction.action": null,
+
+    "azurerm.api_management.policy.xml_content": null,
+    "azurerm.api_management.policy.xml_link": null,
```
[the remark this answers is not in the document]
Comment (jsteinich) [reply packages/cdktf-cli/lib/get/generator/custom-defaults.ts:330]: One of the tickets mentioned the go schema type which causes the issue. I just searched for that type in two of the official GitHub organizations and followed through the go code to map to resources.
I certainly wasn't expecting to find so many instances of it.

[the remark this answers is not in the document]
Comment (skorfmann) [reply packages/cdktf-cli/lib/get/generator/custom-defaults.ts:330]: so, that's nothing we can infer directly from the schema, correct?

[the remark this answers is not in the document]
Comment (jsteinich) [reply packages/cdktf-cli/lib/get/generator/custom-defaults.ts:330]: Correct. The json schema doesn't include everything from the go schema.

**Pull request merged**